src/java.base/share/classes/java/lang/ClassValue.java

```diff
@@ -347,16 +367,14 @@ boolean isLive() {
             Version<T> v = version();
             if (v == null)  return false;
             if (v.isLive())  return true;
-            clear();
+            version.clear();
             return false;
         }
         Entry<T> refreshVersion(Version<T> v2) {
-            assertNotPromise();
-            @SuppressWarnings("unchecked")  // if !isPromise, type is T
-            Entry<T> e2 = new Entry<>(v2, (T) value);
-            clear();
-            // value = null -- caller must drop
-            return e2;
+            if (!version.refersTo(v2)) {
+                return new Entry<>(v2, value);
+            }
+            return this;
```

Comment:
The following reads slightly better (to me):

```suggestion
            return version.refersTo(v2) ? this : new Entry<>(v2, value);
```